addons/website_sale/controllers/main.py

```diff
@@ -388,6 +388,86 @@ def old_product(self, product, category='', search='', **kwargs):
         # Compatibility pre-v14
         return request.redirect(_build_url_w_params("/shop/%s" % slug(product), request.params), code=301)
 
+    def _get_product_carousel_view(self, product_product=None, product_template=None):
+        if product_product:
+            return request.env['ir.ui.view']._render_template('website_sale.shop_product_carousel', values={
+                'product': request.env['product.template'].browse(product_template),
+                'product_variant': product_product,
+            })
+        return request.env['ir.ui.view']._render_template('website_sale.shop_product_carousel', values={
+            'product': product_template,
+        })
+
+    @http.route(['/shop/product/extra-images'], type='json', auth='user', website=True)
+    def add_product_images(self, images, product_product_id, product_template_id, combination_ids=None):
+        """
+        Turns a list of image ids refering to ir.attachments to product.images,
+        links all of them to product.
+        :raises NotFound : If the user is not allowed to access Attachment model
+        """
+
+        if not (request.env.user.has_group('website.group_website_designer') or request.env.user.has_group(
+                'website.group_website_publisher')):
+            raise NotFound()
+
+        image_ids = request.env["ir.attachment"].browse(i['id'] for i in images)
+        image_create_data = [Command.create({
+                    'name': image.name,                          # Images uploaded from url do not have any datas. This recovers them manually
+                    'image_1920': image.datas if image.datas else request.env['ir.qweb.field.image'].load_remote_url(image.url),
+                }) for image in image_ids]
+
+        product_product = request.env['product.product'].browse(int(product_product_id)) if product_product_id else False
+        product_template = request.env['product.template'].browse(int(product_template_id)) if product_template_id else False
+
+        if product_product and not product_template:
+            product_template = product_product.product_tmpl_id
+
+        if product_product and product_template:
+            if product_template.has_configurable_attributes:
+                product_product.write({
+                    'product_variant_image_ids': image_create_data
+                })
+            else:
+                product_template.write({
+                    'product_template_image_ids': image_create_data
+                })
+
+        elif product_template.has_dynamic_attributes():
+            combination = request.env['product.template.attribute.value'].browse(combination_ids)
+            product_product = product_template._get_variant_for_combination(combination)
```

Comment:
Can this really return something in this case? Isn't it guaranteed to not exist if `product_product` is falsy?

Reply:
For some reason, creating a dynamic attribute removes all product.products and only leaves us with the template (At least at first, before creating the variants). Correct me if I'm wrong, but that's why that code is here